Canvas Quizzes Integration

Akshat-Kalra opened the pull request:
# Description

Canvas Quizzes now supported! Also includes 4 levels of access-control, so that instructors can choose how much information from the quiz they want the chatbot to have access to.

TODO:
- [ ] Add integration tests 
- [ ] Edit some prompt on the chatbot side so that it is able to provide quiz answers for those quizzes with FULL ACCESS.
- [ ] Test for mobile view

Preview:

<img width="1793" height="490" alt="image" src="https://github.com/user-attachments/assets/8cf88335-e57e-4931-8e91-e5eef5c37caf" />

<img width="1915" height="1502" alt="image" src="https://github.com/user-attachments/assets/630fad22-a7ad-4457-be74-1e5255c827d2" />




Closes #343

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] This requires a run of `yarn install`
- [ ] This change requires an addition/change to the production .env variables. These changes are below:
- [ ] This change requires developers to add new .env variables. The file and variables needed are below:
- [ ] This change requires a database query to update old data on production. This query is below:


# How Has This Been Tested?

manually for now but will be adding integration tests


# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code where needed 
- [ ] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing tests pass *locally* with my changes
- [ ] Any work that this PR is dependent on has been merged into the main branch
- [ ] Any UI changes have been checked to work on desktop, tablet, and mobile

Akshat-Kalra commented:
@bhunt02 This is my PR for Canvas Quizzes, I've added you as a reviewer, though it's not ready to be merged yet since I've idenfied some TODO's  that I've yet to do.

Though I do have some questions:

Sometimes the quiz question and answer options are in different chunks like

Chunk 1:
`1. This is a question`

Chunk 2:
``` 
Answer Options:
1.  
2.
```

Wouldn’t that affect how specific the chatbot can be when users ask about a question and its options? Is there a way for it to keep track of the sequence?

Akshat-Kalra commented:
@bhunt02

Another question:

Right now helpme is denying to help me with the quiz questions and answers even though it has access to it, which implies there might be a prompt on the chatbot side that is preventing it. So is it okay if I tweak that prompt so that it works for quizzes.

This is the screenshot from my chat for example.
<img width="682" height="715" alt="image" src="https://github.com/user-attachments/assets/3837332a-831b-4218-9b24-9f357c55e8cd" />

bhunt02 commented:
I'll give this a thorough review as soon as I can - with the semester beginning it might not look too good on the timeline.
W.R.T. your questions:
1. If the chunks are too big to contain the full question, they're going to split them and often in inopportune ways. No way around it but to have shorter questions, unless it's erroneously splitting small questions into multiple things. This is purely a problem of the fact chunks can't be too big due to the embedding model's limited context window (and we've got things configured so that chunks are small to ideally contain related information exclusively)
2. Yeah. We've got things set up so that the chatbot will not give direct answers if those materials are included in the source. I don't know if I like the idea of changing it up so that the chatbot will give direct answers by default - and majority of our users won't, either. They want to just upload their documents and forget about it, and I don't blame them. If they find out that our chatbot is giving out assignment answers, we'll be in for a world of trouble. Adam can give some input here, but this is sort of a barrier I'm not willing to budge on.

AdamFipke commented:
@Akshat-Kalra Huh interesting, with regards to 2, I don't recall there being anything specific in any of our prompts that tells the chatbot to not give direct answers if those materials are included in the source. Our system prompt is pretty minimal, from as far as I know.

I know the default course prompt mentions something like "don't provide answers, instead provide hints," though professors can adjust this under Course Settings -> Chatbot Settings -> Open Settings. My guess is you should just be able to adjust the prompt here and remove that sentence

AdamFipke commented on packages/server/src/lmsIntegration/lmsIntegration.adapter.ts:
```diff
@@ -443,6 +451,76 @@ class CanvasLMSAdapter extends ImplementedLMSAdapter {
+    status: LMSApiResponseStatus;
+    quizzes: LMSQuiz[];
+  }> {
+    const { status, data } = await this.GetPaginated(
+      `courses/${this.integration.apiCourseId}/quizzes`,
+    );
+
+    if (status != LMSApiResponseStatus.Success) return { status, quizzes: [] };
+
+    const quizzes: LMSQuiz[] = [];
+
+    for (const quiz of data.filter((q: any) => q.published)) {
+      const { status: quizStatus, data: quizData } = await this.Get(
+        `courses/${this.integration.apiCourseId}/quizzes/${quiz.id}?include[]=questions`,
+      );
```
One worry I have with this is I think it blindly trusts the data returned from Canvas to always be of the correct type, which may be bad if canvas were to change their API or get compromised. It's unlikely but still is potential attack surface area (since ideally we trust no external systems). If LMSQuiz is a class with [class-validator](https://github.com/typestack/class-validator) decorators (e.g. `@IsDate` or `@IsInt`) you can use class-validator's validate() to validate the object and make sure it follows the correct format (though I will note that LMSQuiz is a type not a class, so you will probably want to make a class-equivalent for this case).

This may also be an issue with the other endpoint calls, so maybe this could be saved for a future issue, but I figured it might be good to mention

Akshat-Kalra commented on packages/frontend/app/(dashboard)/course/[cid]/(settings)/settings/lms_integrations/components/QuizContentPreviewModal.tsx:
```diff
@@ -0,0 +1,159 @@
+        </Text>
+      </div>
+
+      {selectedAccessLevel !== currentAccessLevel && (
+        <Alert
+          message="Preview Mode"
+          description={`This shows what the chatbot would see if access level was changed to "${ACCESS_LEVEL_LABELS[selectedAccessLevel]}". Current level is "${ACCESS_LEVEL_LABELS[currentAccessLevel]}".`}
+          type="info"
+          showIcon
+          style={{ marginBottom: 16 }}
+        />
+      )}
```
Oh! there is an alert if you are not on the current access level. Is this good enough, if not I can add the alert for current access level aswell

<img width="725" height="483" alt="image" src="https://github.com/user-attachments/assets/d34edd4f-64fd-47d5-9662-cdc10233dffd" />

FabianLCH replied:
Oh no yeah, I more so imagined it as a quick, hard-to-miss visual queue that reminds users the access level selected in the preview modal at the given modal is the currently configured access level. I thought it may be slightly better than having the preview mode alert disappear e.g. when someone switches from a different once back to the currently selected access level

Akshat-Kalra commented on packages/server/src/lmsIntegration/lmsIntegration.controller.ts:
```diff
@@ -605,6 +755,48 @@ export class LMSIntegrationController {
+      );
+    }
+
+    try {
+      const result = await this.integrationService.updateQuizAccessLevel(
```
if the quiz resource is disabled, its going to show up like this, so they won't hv the UI element that triggers this endpoint.

<img width="1059" height="955" alt="image" src="https://github.com/user-attachments/assets/6e178e75-0cac-47da-aa0f-b7a50327e3a0" />

bhunt02 replied:
it's still a good idea to make sure this endpoint is protected in that way, frontend protections only exist as a means to inform the user of system state and prevent unwanted action. backend protections are the real deal and prevent users from screwing our end of things up

Akshat-Kalra commented on packages/server/src/lmsIntegration/lmsIntegration.service.ts:
```diff
@@ -1267,4 +1367,128 @@ export class LMSIntegrationService {
+        ) {
+          if (q.neutral_comments) {
+            content += `\n   General comments: ${q.neutral_comments}`;
+          }
```
Resolved!

Akshat-Kalra commented on packages/server/src/lmsIntegration/lmsIntegration.service.ts:
```diff
@@ -1267,4 +1367,128 @@ export class LMSIntegrationService {
+
+        if (accessLevel === LMSQuizAccessLevel.FULL_ACCESS && q.answers) {
+          content += '\n   Answer options:';
+          q.answers.forEach((a: any) => {
+            const marker = a.weight > 0 ? ' [CORRECT]' : '';
+            content += `\n   - ${a.text}${marker}`;
+          });
```
Technically yeah, I'll add it right now

Akshat-Kalra replied:
Though on a side note, this is the documentation for the canvas api `/questions` endpoint and it mentions nothing about the `_html` but it shows up in my returned object for question comments but not answer comments

bhunt02 commented:
<img width="503" height="185" alt="image" src="https://github.com/user-attachments/assets/709a8880-9367-4c1e-810f-78da11dd2370" />
this is a bit concerning - why is the description not attached (this is a chunk preview right?)

bhunt02 commented on packages/server/src/lmsIntegration/lmsIntegration.adapter.ts:
```diff
@@ -443,6 +451,77 @@ class CanvasLMSAdapter extends ImplementedLMSAdapter {
+    status: LMSApiResponseStatus;
+    quizzes: LMSQuiz[];
+  }> {
+    const { status, data } = await this.GetPaginated(
```
!! make sure you filter out the unpublished ones !!

bhunt02 reviewed (changes requested):
at first I was sort of uneasy about the complexity of the interface (instructors tend to not want to interact too much with buttons and toggles), but it's definitely required for this particular document type ,and if they're determined enough to use it, that's what they have to do!

it looks really well-thought-out - i just want 
just want some small changes - those previews are going to need to use the dangerouslySetInnerHtml property on the element so that it actually renders the tags - it's kind of ugly to see a bunch of HTML tags :( ALSO - it's VERY important that we do not pull any unpublished documents.

bhunt02 commented:
<img width="955" height="467" alt="image" src="https://github.com/user-attachments/assets/69453e46-9325-4a14-9c1e-cc0cc082272a" />
question: why not include the prefix in the content that will be sent? might give a better picture of what information will be stored :)

bhunt02 commented on packages/server/src/lmsIntegration/lmsIntegration.controller.ts:
```diff
@@ -332,6 +336,147 @@ export class LMSIntegrationController {
+          where: { courseId, id: In(body.quizIds) },
+        });
+
+        for (const quiz of enabledQuizzes) {
```
should *this* whole operation fail if one quiz fails to sync?

bhunt02 commented on packages/server/src/lmsIntegration/lmsIntegration.adapter.ts:
```diff
@@ -443,6 +451,77 @@ class CanvasLMSAdapter extends ImplementedLMSAdapter {
+
+    const quizzes: LMSQuiz[] = [];
+
+    for (const quiz of data.filter((q: any) => q.published)) {
```
me when i cant read

bhunt02 reviewed (changes requested):
some concerns but overall really good work

bhunt02 reviewed (approved):
I'm gonna leave an approval now since my concerns were addressed, and it looks like you were quite thorough with all this. Just make sure to make the tests pass!

FabianLCH reviewed (approved):
I'll go ahead and leave my approval too since the work looks good and most of my comments were addressed. Looks like tests are failing because of a mistyped variable name, so after fixing if all tests pass we should be good to go

AdamFipke commented:
Don't worry about a review from me. I'm still not super familiar with the LMS stuff and I don't really have a way to manually test it, so if fabian and bridgette are happy with it it should be good to merge!

Akshat-Kalra commented:
> I'm gonna leave an approval now since my concerns were addressed, and it looks like you were quite thorough with all this. Just make sure to make the tests pass!

Thanks @bhunt02 for the review!!

Akshat-Kalra commented:
> I'll go ahead and leave my approval too since the work looks good and most of my comments were addressed. Looks like tests are failing because of a mistyped variable name, so after fixing if all tests pass we should be good to go

Thanks @FabianLCH for the review; I've addressed your suggestions about backend endpoint protections and made necessary changes! Once these tests pass, I'll go ahead and merge this branch.
